sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -1243,9 +1248,12 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::ExtractOrCreateCurlCo
   std::string const connectionKey = GetConnectionKey(host, options);
 
   {
+    decltype(CurlConnectionPool::g_curlConnectionPool
+                 .ConnectionPoolIndex)::mapped_type connectionsToBeReset;
+
     // Critical section. Needs to own ConnectionPoolMutex before executing
     // Lock mutex to access connection pool. mutex is unlock as soon as lock is out of scope
-    std::lock_guard<std::mutex> lock(CurlConnectionPool::ConnectionPoolMutex);
+    std::unique_lock<std::mutex> lock(CurlConnectionPool::ConnectionPoolMutex);
```

Comment:
Do you think `ConnectionPoolMutex` would benefit from being `std::shared_timed_mutex`, so in some cases we can do `std::shared_lock<std::shared_timed_mutex>`, and `std::unique_lock<std::shared_timed_mutex>` in others?

Reply:
I don't see how we can benefit from that.

`std::shared_timed_mutex` provides two more functionalities than `std::mutex`, timeout and sharable ownership.

For sharable ownership, currently there are three operations that need to be protected with mutex, insert a connection into pool, fetch a connection from the pool and clean the connection pool. All of these three operations require exclusive ownership because we need to modify the index. So sharable ownership won't help here.

For timeout, I don't think there's anything we can do if we've waited on the mutex for too long time.